src/bootstrap/test.rs

```diff
@@ -704,6 +704,48 @@ impl Step for RustdocUi {
     }
 }
 
+#[derive(Debug, Copy, Clone, Hash, PartialEq, Eq)]
+pub struct ErrorCodesCheck {
+    pub compiler: Compiler,
+}
+
+impl Step for ErrorCodesCheck {
+    type Output = ();
+    const DEFAULT: bool = true;
+    const ONLY_HOSTS: bool = true;
+
+    fn should_run(run: ShouldRun<'_>) -> ShouldRun<'_> {
+        run.path("src/tools/error-codes")
+    }
+
+    fn make_run(run: RunConfig<'_>) {
+        let compiler = run.builder.compiler(run.builder.top_stage, run.host);
+
+        run.builder.ensure(ErrorCodesCheck { compiler });
+    }
+
+    fn run(self, builder: &Builder<'_>) {
+        let rustdoc = builder.out.join("bootstrap/debug/rustdoc");
+
+        builder.ensure(compile::Std { compiler: self.compiler, target: self.compiler.host });
+
+        let mut cmd = builder.tool_cmd(Tool::ErrorCodesCheck);
+        cmd.arg(rustdoc.to_str().unwrap())
+            .arg(builder.src.join("src/librustc_error_codes/error_codes").to_str().unwrap())
+            .env("RUSTC_STAGE", self.compiler.stage.to_string())
+            .env("RUSTC_SYSROOT", builder.sysroot(self.compiler))
+            .env("RUSTDOC_LIBDIR", builder.sysroot_libdir(self.compiler, self.compiler.host))
+            .env("CFG_RELEASE_CHANNEL", &builder.config.channel)
+            .env("RUSTDOC_REAL", builder.rustdoc(self.compiler))
+            .env("RUSTDOC_CRATE_VERSION", builder.rust_version())
+            .env("RUSTC_BOOTSTRAP", "1");
```

Comment:
Do you actually need most of these? I imagine that you could directly pass `builder.rustdoc(compiler)` as the argument to the tool and omit ~all of these variables.